### PR TITLE
Use the new Decorator API for highlights

### PR DESCRIPTION
**Pull request by mickael-menu**
[See the specification of the Decorator API](https://github.com/readium/architecture/pull/160)

Depends on https://github.com/readium/r2-navigator-kotlin/pull/214

* Remove the `EpubActivity` since we don't need to inherit from `R2EpubActivity` anymore.
* Add an underline style (and selection menu item).
* Add a custom HTML decoration template for a side mark icon, as an example.
* Rewrite the `Highlight` DB model.
* Move the highlighting code at the `BaseReaderFragment` level, as it's not specific to EPUB anymore.

### CODE REVIEW

File: r2-testapp/src/main/java/org/readium/r2/testapp/outline/HighlightsFragment.kt [hunk 1]
```diff
@@ -58,18 +61,16 @@ class HighlightsFragment : Fragment() {
...
 
-        highlightAdapter = HighlightAdapter(publication, onDeleteHighlightRequested = { highlight -> viewModel.deleteHighlightByHighlightId(highlight.highlightId) }, onHighlightSelectedRequested = { highlight -> onHighlightSelected(highlight) })
+        highlightAdapter = HighlightAdapter(publication, onDeleteHighlightRequested = { highlight -> viewModel.deleteHighlight(highlight.id) }, onHighlightSelectedRequested = { highlight -> onHighlightSelected(highlight) })
         binding.listView.apply {
             setHasFixedSize(true)
             layoutManager = LinearLayoutManager(requireContext())
             adapter = highlightAdapter
         }
 
-        val comparator: Comparator<Highlight> = compareBy({ it.resourceIndex }, { it.locator.locations.progression })
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/outline/HighlightsFragment.kt:68]: Sorting is done in the SQL query itself now.

File: r2-testapp/src/main/java/org/readium/r2/testapp/reader/BaseReaderFragment.kt [hunk 2]
```diff
@@ -52,16 +65,56 @@ abstract class BaseReaderFragment : Fragment() {
...
+                }
+            )
+
+            model.highlightDecorations
+                .onEach { navigator.applyDecorations(it, "highlights") }
+                .launchIn(viewScope)
+
+            model.searchDecorations
+                .onEach { navigator.applyDecorations(it, "search") }
+                .launchIn(viewScope)
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/BaseReaderFragment.kt:109]: 👀

File: r2-testapp/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt [hunk 3]
```diff
@@ -63,7 +83,16 @@ class EpubReaderFragment : VisualReaderFragment(), EpubNavigatorFragment.Listene
...
-            EpubNavigatorFragment.createFactory(publication, baseUrl, model.initialLocation, this)
+            EpubNavigatorFragment.createFactory(
+                publication = publication,
+                baseUrl = baseUrl,
+                initialLocator = model.initialLocation,
+                listener = this,
+                config = EpubNavigatorFragment.Configuration().apply {
+                    // Register the HTML template for our custom [DecorationStyleAnnotationMark].
+                    decorationTemplates[DecorationStyleAnnotationMark::class] = annotationMarkTemplate(activity)
+                }
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt:94]: 👀

File: r2-testapp/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt [hunk 4]
```diff
@@ -268,4 +325,46 @@ class EpubReaderFragment : VisualReaderFragment(), EpubNavigatorFragment.Listene
...
+}
+
+/**
+ * Example of an HTML template for a custom Decoration Style.
+ *
+ * This one will display a tinted "pen" icon in the page margin to show that a highlight has an
+ * associated note.
+ */
+@OptIn(ExperimentalDecorator::class)
+private fun annotationMarkTemplate(context: Context, @ColorInt defaultTint: Int = Color.YELLOW): HtmlDecorationTemplate {
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt:337]: 👀

Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt:337]: 👀

File: r2-testapp/src/main/java/org/readium/r2/testapp/reader/ReaderViewModel.kt [hunk 5]
```diff
@@ -67,27 +68,83 @@ class ReaderViewModel(context: Context, arguments: ReaderContract.Input) : ViewM
...
+                }
+            ),
+            // Additional page margin icon decoration, if the highlight has an associated note.
+            annotation.takeIf { it.isNotEmpty() }?.let {
+                createDecoration(
+                    idSuffix = "annotation",
+                    style = DecorationStyleAnnotationMark(tint = tint),
+                )
+            }
+        )
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/ReaderViewModel.kt:128]: 👀

File: r2-testapp/src/main/java/org/readium/r2/testapp/reader/ReaderViewModel.kt [hunk 6]
```diff
@@ -111,6 +168,24 @@ class ReaderViewModel(context: Context, arguments: ReaderContract.Input) : ViewM
...
+                Decoration(
+                    // The index in the search result list is a suitable Decoration ID, as long as
+                    // we clear the search decorations between two searches.
+                    id = index.toString(),
+                    locator = locator,
+                    style = Decoration.Style.Underline(tint = Color.RED)
+                )
+            }
+        }
+    }
```
Comment (mickael-menu) [inline r2-testapp/src/main/java/org/readium/r2/testapp/reader/ReaderViewModel.kt:187]: 👀

**Review by qnga — APPROVED** [review]
Excepted that CI complains for some reason, it looks good!

**Comment by mickael-menu** [comment]
@qnga It uses the latest `develop` commits for the R2 modules, so the build should solve itself once the other PR is merged.

**Pull request merged**